core/server/worker/src/main/java/alluxio/worker/block/UfsIOManager.java

```diff
@@ -0,0 +1,298 @@
+    return ufsBytesReadThroughput.getOneMinuteRate();
+  }
+
+  /**
+   * read from ufs.
+   * 
+   * @param blockId block id
+   * @param blockSize block size
+   * @param ufsPath ufs path
+   * @param offset offset to read
+   * @param length length to read
+   * @param positionShort is position short or not
+   * @param user user tag
+   * @return content read
+   * @throws ResourceExhaustedException any exception from ufs
+   */
+  public CompletableFuture<byte[]> read(long blockId, long blockSize, String ufsPath, long offset,
+      long length, boolean positionShort, String user) throws ResourceExhaustedException {
+    CompletableFuture<byte[]> future = new CompletableFuture<>();
+    if (mReadQueue.size() >= READ_CAPACITY) {
+      throw new ResourceExhaustedException(format("UFS read at capacity, try later or increase %s",
+          PropertyKey.UNDERFS_IO_READ_QUEUE_CAPACITY.getName()));
+    }
+    mReadQueue.add(
+        new ReadTask(blockId, blockSize, ufsPath, offset, length, positionShort, user, future));
+    return future;
+  }
+
+  private class ReadTask implements Runnable {
+    private final long mOffset;
+    private final long mLength;
+    private final boolean mIsPositionShort;
+    private final CompletableFuture<byte[]> mFuture;
+    private final long mBlockId;
+    private final String mUfsPath;
+    private final String mUser;
+    private final long mBlockSize;
+
+    private ReadTask(long blockId, long blockSize, String ufsPath, long offset, long length,
+        boolean positionShort, String user, CompletableFuture<byte[]> future) {
+      mUser = user;
+      mBlockId = blockId;
+      mBlockSize = blockSize;
+      mUfsPath = ufsPath;
+      mOffset = offset;
+      mLength = length;
+      mIsPositionShort = positionShort;
+      mFuture = future;
+    }
+
+    public String getUser() {
+      return mUser;
+    }
+
+    public void run() {
+      try {
+        byte[] buffer = readInternal();
+        mFuture.complete(buffer);
+      } catch (IOException e) {
+        mFuture.completeExceptionally(e);
+      }
+    }
+
+    public byte[] readInternal() throws IOException {
+      long bytesToRead = Math.min(mLength, mBlockSize - mOffset);
+      if (bytesToRead <= 0) {
+        return new byte[0];
+      }
+      UnderFileSystem ufs = mUfsClient.acquireUfsResource().get();
+      long blockStart = BlockId.getSequenceNumber(mBlockId) * mBlockSize;
+      byte[] data = new byte[(int) bytesToRead];
+      int bytesRead = 0;
+      InputStream inStream = null;
+      try {
+        inStream = mUfsInstreamCache.acquire(ufs, mUfsPath, IdUtils.fileIdFromBlockId(mBlockId),
+            OpenOptions.defaults().setOffset(blockStart + mOffset)
+                .setPositionShort(mIsPositionShort));
+        while (bytesRead < bytesToRead) {
+          int read;
+          read = inStream.read(data, bytesRead, (int) (bytesToRead - bytesRead));
```

Comment:
In what situation will the number of bytes read be less than `bytesToRead`? Also, the cached input stream implementation is weird. the offset here is offset relative to the offset when the cached stream is acquired? This makes the code very hard to read / reason. Unless we can prove that the cached input stream has performance benefit, we should probably not use it.

Reply:
My personal feeling of relative offset is fine. let's keep it for now